clients/src/main/java/org/apache/kafka/common/log/remote/metadata/storage/ConsumerTask.java

```diff
@@ -209,29 +225,30 @@ public void syncCommittedDataAndOffsets(boolean forceSync) {
         }
     }
 
-    public void reassignForPartitions(Set<TopicPartition> partitions) {
-        Objects.requireNonNull(partitions, "partitions can not be null");
+    public void addAssignmentsForPartitions(Set<TopicPartition> updatedPartitions) {
+        Objects.requireNonNull(updatedPartitions, "partitions can not be null");
 
-        log.info("Reassigning for user partitions {}", partitions);
-        synchronized (lock) {
-            // check for the corresponding partitions.
-            Set<Integer> newlyAssignedPartitions = new HashSet<>();
-            for (TopicPartition tp : partitions) {
-                newlyAssignedPartitions.add(remoteLogSegmentMetadataUpdater.metadataPartitionFor(tp));
-            }
-            reassignedTopicPartitions = new HashSet<>(partitions);
-            assignedMetaPartitions = Collections.unmodifiableSet(newlyAssignedPartitions);
-            log.info("Newly assigned metadata partitions {}", assignedMetaPartitions);
-            reassign = true;
-
-            lock.notifyAll();
-        }
+        log.info("Reassigning for user partitions {}", updatedPartitions);
+        updateAssignmentsForPartitions(updatedPartitions, Collections.emptySet());
     }
 
     public void removeAssignmentsForPartitions(Set<TopicPartition> partitions) {
+        updateAssignmentsForPartitions(Collections.emptySet(), partitions);
+    }
+
+    private void updateAssignmentsForPartitions(Set<TopicPartition> addedPartitions,
+                                                Set<TopicPartition> removedPartitions) {
+        Objects.requireNonNull(addedPartitions,"addedPartitions must not be null");
+        Objects.requireNonNull(removedPartitions,"removedPartitions must not be null");
+
+        if(addedPartitions.isEmpty() && removedPartitions.isEmpty()) {
+            return;
+        }
+
         synchronized (lock) {
```

Comment:
ConsumerTask.run() holds this lock when polling the metadata topic. If there is no new messages in the metadata topic, it can block up to 30 seconds.

updateAssignmentsForPartitions is called in Kafka request handler threads. This lock may block request handler thread for 30 seconds.

We actually do not need to hold the lock while polling the metadata topic.

Reply:
`lock` is not taken when it polls messages from the metadata topic using `consumer.poll()`. It leaves the lock before that.